// fastjsonweb/lib/cookie.js
var  EventEmitter = require('events').EventEmitter;
function Cookie(req,res){
	/*
	this.parseCookie= function(){
		console.log("run");
		var arr= [];
		var obj = {};
		if(req.headers.cookie){
			arr = req.headers.cookie.split(";");
		}
		for(index in arr){
			var parts = arr[index].split('=');
			obj[parts[0].trim()] = (parts[1]||'').trim();
		}
		console.log(obj);
		console.log("runed");
		return obj;
	}
	*/
	//解析cookie数据
	this.Cookies = (function(){
		//var  emitter = new EventEmitter();
		//emitter.emit("hello","huyinghuan");
		console.log("run");
		var arr= [];
		var obj = {};
		if(req.headers.cookie){
			arr = req.headers.cookie.split(";");
		}
		for(index in arr){
			var parts = arr[index].split('=');
			obj[parts[0].trim()] = (parts[1]||'').trim();
		}
		console.log(obj);
		console.log("runed");
		return obj;
	})();
	//获取cookie
	this.getValue = function(key){
		console.log("get"+JSON.stringify(this.Cookies))
		return this.Cookies ? (this.Cookies[key]||undefined) :undefined;
	};
	//设置Cookie
	this.setValue= function(key,value){
		if(typeof(value) !== "string"){
			value = JSON.stringify(value);
		}
		this.Cookies[key] = value;
		console.log("set"+JSON.stringify(this.Cookies))
	};
	//设置获取sid
	this.getSessionId= function(){
		return this.Cookies["sid"]||undefined;
	},
	this.setSessionId= function(sid){
		this.Cookies["sid"] = sid;
	}
	//写入输出流
	this.flush = function(){
		console.log("执行 cookie flush")
		console.log(this.Cookies);
		var arr = [];
		for(key in this.Cookies){
			arr.push(key+"="+this.Cookies[key]);
		}
		return arr;
		console.log(arr);
		res.setHeader('Set-Cookie',arr);
	}
}
module.exports = Cookie;
