// fastjsonweb/lib/globelsetting.js
var definedconfigure ={
	sessiontimeout : 1000*60*30, //session legal time
	static : "/public",
	smartrouter:false
}
//set session legal time
function sessiontimeout(time){
	if(typeof time === "number"){
		definedconfigure["sessiontimeout"] = time*1000;
	}else if(typeof time === "object"){
		 var hours = typeof time["hours"] === "number" ? time["hours"] : 0;
		 var minutes = typeof time["minutes"] === "number" ? time["minute"] : 0;
		 var seconds = typeof time["seconds"] === "number" ? time["seconds"] : 0;
		 var totaltime = hours*60*60+minutes*60+seconds;
		 if(totaltime > 0){
			definedconfigure["sessiontimeout"] = totaltime*1000;
		 }else{
			console.log(" session time can't less than 0,use the default time 30 minutes");
		 }
	}else{
		console.log("session time defined the error,use the default time 30 minutes");
	}
}
// set static resource direct
function staticresourcedirect(path){
	definedconfigure["static"] = typeof path === "string" ? path : "/public";
}

function smartrouter(b){
	if(typeof b === "boolean")
		definedconfigure.smartrouter = b;
}

exports.setting = definedconfigure;

exports.setConfigure = function(setting){
	if(typeof setting !== "object") return;
	sessiontimeout(setting["sessiontimeout"]);
	staticresourcedirect(setting["static"]);
	smartrouter(setting["smartrouter"]);
}
