// fastjsonweb/lib/NewReponse.js
var mime = require('./mime').types;
function NewResponse(response){
	this.data = {
		type:"JSON",
		header:{},
		status:200
		/*content: String or binary,default string
		isfile: true or false ,defualt false,
		filesetting: { 
			download: true or false,  false:open the file on brower,true,download this file,default false
			filename: string,
			filetype: string  the file type see "./mime.js"
		}
		*/
	};
	this.setReponseType = function(type){
		this.data.type = type;
	};
	this.setHeader = function(key,value){
		this.data.header[key] = value;
	};
	this.setStatus = function(code){
		this.data.status = (typeof code === "number" || +code == code )? +code : 200;
	}
	this.write = function(content,isfile,filesetting){
		if(!!isfile)
			this.data
		this.data.content = content;
	};
	this.end = function(content){
		response.writeHead(this.data.status,{
			'Content-Length':this.data.content.length,
			'Content-Type' : mime[this.data.type]
		});
		response.write((content ? content : this.data.content));
		response.end(); 
	};
	this.called = function(name,arg){
		response[name].apply(undefined,arg);
	};
	this.getResponse = function(){
		return response;
	}
}
module.exports = NewResponse;
