// fastjsonweb/lib/property_injection.js
/**
    =========================================
    修改urlParameterToObject函数命名 改为parameterToObject
*/
//判断一个对象是否为空 ，对象为null则ture，否则false
function isEmptyObject(obj){
  if(!obj){
    return true;
  }
  for(x in obj){
    return false;
  }
  return true;
}

//a.b.c.d = 123; {b:{c:{d:"123"}}}

function parseObjFromArray(obj,value,arr,i){
  if(i == 0){
    return obj;
  }else{
    obj = {};
  }
  var property = arr[i-1];
  obj[property]  = value;
  return parseObjFromArray(obj,obj,arr,i-1);
}

//求两个对象并集
function unionObject(firstObj,secondObj){
   var unionObj = {};
   if(!secondObj || isEmptyObject(secondObj)){
      return firstObj;
   }
   for(x in firstObj){
      for(y in secondObj){
         if(firstObj[x] === secondObj[y]){
           unionObj[x] = secondObj[x];
           continue;
         }
         
         if(x != y){
           unionObj[x] = firstObj[x];
           if(!unionObj[y]){
            unionObj[y] = secondObj[y];
           }
           continue;
         }
         var type = typeof secondObj[x] ;
         
         if(type == 'string'){
             unionObj[x] = secondObj[x];
             continue;
         }
         
         if(type == 'function'){
            unionObj[x] = secondObj[x];
            continue;
         }
         
         if(type == 'object'){
            var prop  = x;
            var ob = unionObject(firstObj[x],secondObj[x]);
            unionObj[prop] = ob;
            continue;
         }
      }
   }
   return unionObj;
}

//求一组对象的并集
function unionObjectFromArray(arrObj){
  if(!arrObj || arrObj.length == 0){
    return {};
  }
  if(arrObj.length == 1){
    return arrObj[0];
  }
  var length = arrObj.length;

  var objs = new Array();
  for(i = 0;i < length - 1;i = i+2){
    objs.push(unionObject(arrObj[i],arrObj[i+1]));
  }
  if(length%2 != 0){
    objs.push(arrObj[length-1]);
  }
  return unionObjectFromArray(objs);
}

function stringToObject(str,value){
   var strtmp = str.replace(/\.+/g,'\.');
   var errMsg;
   //判断 ‘.’ 是否重复或过多
   if(strtmp != str){
      errMsg = "ERROR: This is not property betwen two '.'  at \"" + str+"="+value+"\"";
      console.log(typeof errMsg);
      console.log(errMsg);
      return;
   }
   var index = str.indexOf('.');
   //判断 ‘.’是否为位于首位或末位
   if(index == 0 || index ==(str.length - 1)){
      errMsg = "ERROR: This is no object before '.' or property after '.' at \"" + str+"\"";
      console.log(typeof errMsg);
      console.log(errMsg);
      return;
   }
   var arr = str.split('\.');
   var length = arr.length;
   var object = parseObjFromArray({},value,arr,length);
   return object;
}

function parameterToObject(urlParameter){
  var propertyObj   = {};
  if(isEmptyObject(urlParameter)){
    return propertyObj;
  }
  var arrObject = new Array();
  var parameterValue ='';
  for(parameterName in urlParameter){
      var index = parameterName.indexOf('\.');
      parameterValue = urlParameter[parameterName];
      if( index == -1){
         propertyObj[parameterName] = parameterValue;
         continue;
       }
       arrObject.push(stringToObject(parameterName,parameterValue));
  }
  
  arrObject.push(propertyObj);
  return unionObjectFromArray(arrObject);
}

exports.parameterToObject = parameterToObject;
exports.isEmptyObject = isEmptyObject;
exports.unionObject = unionObject;
